gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergDataset.java

```diff
@@ -188,11 +215,19 @@ protected Map<Path, FileStatus> getFilePathsToFileStatus(FileSystem targetFs, Co
           this.getFileSetId(), currentSnapshotOverview.getSnapshotId(),
           currentSnapshotOverview.getManifestListPath(),
           currentSnapshotOverview.getMetadataPath().orElse("<<ERROR: MISSING!>>"));
-      return Maps.newHashMap();
+      TableMetadata readTimeTableMetadata = currentSnapshotOverview.getTableMetadata().orElseThrow(() -> new RuntimeException(
+          String.format("~%s~ no table metadata for current snapshot '%s' at '%s' with metadata path '%s'",
+              this.getFileSetId(), currentSnapshotOverview.getSnapshotId(),
+              currentSnapshotOverview.getManifestListPath(),
+              currentSnapshotOverview.getMetadataPath().orElse("<<ERROR: MISSING!>>"))));
+      return new GetFilePathsToFileStatusResult(Maps.newHashMap(), readTimeTableMetadata);
     }
+
+    List<TableMetadata> readTimeTableMetadataHolder = Lists.newArrayList(); // expecting exactly one elem
     Iterator<IcebergSnapshotInfo> icebergIncrementalSnapshotInfos = icebergTable.getIncrementalSnapshotInfosIterator();
     Iterator<String> filePathsIterator = Iterators.concat(
         Iterators.transform(icebergIncrementalSnapshotInfos, snapshotInfo -> {
+          snapshotInfo.getTableMetadata().ifPresent(readTimeTableMetadataHolder::add);
           // log each snapshot, for context, in case of `FileNotFoundException` during `FileSystem.getFileStatus()`
           String manListPath = snapshotInfo.getManifestListPath();
           log.info("~{}~ loaded snapshot '{}' at '{}' from metadata path: '{}'", this.getFileSetId(),
```

Comment:
no, mismatch metadata at this point is no problem--in fact it's just part of the src vs. dest diff calc.  the issue is that diff calc occurs, and then (much) later, when the `IcebergRegisterStep` runs after all of the copying, the state of the source and dest table may have changed since that diff calc.

so classic race condition.  to prevent undefined results: we preserve the metadata used by the diff calc.
* if the source has since changed, we still use what we stored, NOT what's then current, as it wouldn't match the files copied to dest.
* if the dest has since changed, we reject the catalog commit, because to ensure safety the diff calc should re-start again from the beginning.